platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/proof/algorithms/StateProofTreeBuilder.java

```diff
@@ -0,0 +1,306 @@
+     * @return the processed signatures
+     */
+    @NonNull
+    public static List<NodeSignature> processSignatures(@NonNull final Map<NodeId, Signature> unprocessedSignatures) {
+        if (unprocessedSignatures.isEmpty()) {
+            throw new IllegalArgumentException("signatures must not be empty");
+        }
+
+        final List<NodeSignature> signatures = new ArrayList<>(unprocessedSignatures.size());
+        for (final Map.Entry<NodeId, Signature> entry : unprocessedSignatures.entrySet()) {
+            if (entry.getValue() == null) {
+                throw new IllegalArgumentException("signatures are not permitted to contain null values");
+            }
+            signatures.add(new NodeSignature(entry.getKey(), entry.getValue()));
+        }
+
+        Collections.sort(signatures);
+        return signatures;
+    }
+
+    /**
+     * Check if a merkle node is an ancestor of at least one of the payloads. Merkle nodes are considered to be
+     * ancestors to themselves.
+     *
+     * @param potentialAncestor the potential ancestor
+     * @param payloads          the payloads
+     * @return true if the potential ancestor is an ancestor of at least one payload, otherwise false
+     */
+    private static boolean isAncestorOfPayload(
+            @NonNull final MerkleNode potentialAncestor, @NonNull final List<MerkleLeaf> payloads) {
+
+        for (final MerkleLeaf payload : payloads) {
+            if (potentialAncestor.getRoute().isAncestorOf(payload.getRoute())) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Given a merkle root and a list of payloads, extract a list of merkle nodes that are ancestors of at least one of
+     * the payloads. Nodes are returned in reverse post-ordered depth first traversal order.
+     *
+     * @param merkleRoot the merkle root
+     * @param payloads   the payloads
+     * @return a list of merkle nodes that are ancestors of at least one of the payloads
+     * @throws IllegalStateException if the payloads and the merkle tree are inconsistent with each other
+     */
+    private static List<MerkleNode> getMerkleNodesForStateProofTree(
+            @NonNull final MerkleNode merkleRoot, @NonNull final List<MerkleLeaf> payloads) {
+
+        // Note: we use a reverse post-ordered depth first traversal here intentionally. This cause a node's children
+        // to be popped off the top of a stack in left-to-right order (as opposed to a standard DFS traversal,
+        // which will pop them off in right-to-left order).
+
+        final List<MerkleNode> nodes = new ArrayList<>();
+        merkleRoot
+                .treeIterator()
+                .setOrder(MerkleIterationOrder.REVERSE_POST_ORDERED_DEPTH_FIRST)
+                .ignoreNull(true)
+                .setFilter(node -> isAncestorOfPayload(node, payloads))
+                .setDescendantFilter(node -> isAncestorOfPayload(node, payloads))
```

Comment:
Curious why we have two identical state-capturing lambdas set for filter and descendant filter.  Is the filter and descendant filter _intended_ to be the same functionality?